pkg/healthcheck/healthcheck_output.go

```diff
@@ -114,95 +140,147 @@ func PrintChecksResult(wout io.Writer, output string, success bool, warning bool
-			//nolint:gosec
-			plugin := exec.Command(path, args...)
-			var stdout, stderr bytes.Buffer
-			plugin.Stdout = &stdout
-			plugin.Stderr = &stderr
-			plugin.Run()
-			extensionResults, err := parseJSONCheckOutput(stdout.Bytes())
-			spin.Stop()
-			if err != nil {
-				command := fmt.Sprintf("%s %s", path, strings.Join(args, " "))
-				if len(stderr.String()) > 0 {
-					err = errors.New(stderr.String())
-				} else {
-					err = fmt.Errorf("invalid extension check output from \"%s\" (JSON object expected):\n%s\n[%w]", command, stdout.String(), err)
-				}
-				results.Results = append(results.Results, CheckResult{
-					Category:    CategoryID(extensionCmd),
-					Description: fmt.Sprintf("Running: %s", command),
-					Err:         err,
-					HintURL:     HintBaseURL(version.Version) + "extensions",
-				})
-				success = false
+		plugin := utilsexec.Command(extension.path, args...)
+		var stdout, stderr bytes.Buffer
+		plugin.SetStdout(&stdout)
+		plugin.SetStderr(&stderr)
+		plugin.Run()
+		results, err := parseJSONCheckOutput(stdout.Bytes())
+		spin.Stop()
+		if err != nil {
+			success = false
+
+			command := fmt.Sprintf("%s %s", extension.path, strings.Join(args, " "))
+			if len(stderr.String()) > 0 {
+				err = errors.New(stderr.String())
 			} else {
-				results.Results = append(results.Results, extensionResults.Results...)
+				err = fmt.Errorf("invalid extension check output from \"%s\" (JSON object expected):\n%s\n[%w]", command, stdout.String(), err)
+			}
+			_, filename := filepath.Split(extension.path)
+			results = CheckResults{
+				Results: []CheckResult{
+					{
+						Category:    CategoryID(filename),
+						Description: fmt.Sprintf("Running: %s", command),
+						Err:         err,
+						HintURL:     HintBaseURL(version.Version) + "extensions",
+					},
+				},
 			}
 		}
 
-		var extensionSuccess bool
-		extensionSuccess, warning = RunChecks(wout, werr, results, output)
+		extensionSuccess, extensionWarning := RunChecks(wout, werr, results, output)
 		if !extensionSuccess {
 			success = false
 		}
+		if extensionWarning {
+			warning = true
+		}
+	}
+
+	for _, m := range missing {
+		results := CheckResults{
+			Results: []CheckResult{
+				{
+					Category:    CategoryID(m),
+					Description: fmt.Sprintf("Linkerd extension command %s exists", m),
+					Err:         &exec.Error{Name: m, Err: exec.ErrNotFound},
+					HintURL:     HintBaseURL(version.Version) + "extensions",
+					Warning:     true,
+				},
+			},
+		}
+
+		extensionSuccess, extensionWarning := RunChecks(wout, werr, results, output)
+		if !extensionSuccess {
+			success = false
+		}
+		if extensionWarning {
+			warning = true
+		}
 	}
 
 	return success, warning
 }
 
+// FindExtensions searches the path for all linkerd-* executables and returns a
+// slice of check commands, and a slice of missing checks.
+func FindExtensions(pathEnv string, glob Glob, exec utilsexec.Interface, nsLabels []string) ([]Extension, []string) {
```

Comment:
Dumb question: Does this function need to be public? Why?

Reply:
not necessarily? at the moment we're calling it from `cli/cmd` so needs to be public. we could move it down into `pkg/healthcheck`, and call it as part of `RunExtensionsChecks()` (or create a new public function that wraps both).

Reply:
OK, I think what I'm getting at is that I don't actually understand what we view as the scope/public api of `pkg/healthcheck`. Whether we like it or not, this is a public API. We haven't done a great job of formalizing that, but we still want to be slower and more intentional to public API changes.

Is it possible for `findExtensions` to be made a private function in `cli/cmd`?

Basically (and I haven't really digested all of the comments on this PR), my understanding is that we're uncomfortable about the API being introduced. Are we able to derisk that by making a smaller change to the healthcheck api?

Reply:
We can definitely make `findExtensions` a private function in `cli/cmd`.

I think what's important is that it's implemented in a way that we can test different combinations of path, glob, exec, and namespaces, which is doable independent of where it lives and whether it's public or private.

Reply:
One caveat: `findExtensions` returns a `[]healthcheck.Extension`. That struct has a private members that we'll need to make public for `findExtensions` to populate it.

Reply:
OK, and IIUC, we could even mitigate that by running `RunExtensionsCheck` into the CLI package as well... https://github.com/search?q=RunExtensionsCheck&type=code -- at least, there's no code out there in the public that uses this method, afaict.

Though this change is somewhat superficial, I think it will help neutralize concerns about the API so that these internals are less stress to change in the future!